drv/ArrayList.drv

```diff
@@ -662,10 +666,14 @@ public class ARRAY_LIST KEY_GENERIC extends ABSTRACT_LIST KEY_GENERIC implements
 			@Override
 			public void forEachRemaining(final METHOD_ARG_KEY_CONSUMER action) {
 				final KEY_GENERIC_TYPE[] a = ARRAY_LIST.this.a;
-				final int max = to - from;
-				while(pos < max) {
-					action.accept(a[from + (lastReturned = pos++)]);
+				final int from = SubList.this.from, to = SubList.this.to;
+				for (int i = from + pos; i < to; i++) {
+					action.accept(a[i]);
 				}
+
+				final int max = to - from;
+				pos = max;
+				lastReturned = max - 1;
 			}
```

Comment:
The `forEachRemaining` change is meaningless (maybe slower) in the latest benchmark, and I'm not sure if I should keep it. I did this optimization because Java's `ArrayList` uses the same code, and it seems more efficient (less assignment and self-increment).
Benchmark:
```java
@Fork(1)
@State(Scope.Thread)
@Warmup(iterations = 3)
@Measurement(iterations = 5)
@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
public class ForEachRemainingBenchmark {
    @Param({"10", "1000", "100000"})
    private int max;
    private int[] a;

    @Setup
    public void setup() {
        a = new int[max];
        Arrays.setAll(a, IntUnaryOperator.identity());

    }

    private int pos;
    private int last;

    @Benchmark
    public void raw(Blackhole blackhole) {
        pos = max >> 1;
        last = pos - 1;

        while (pos < max) {
            blackhole.consume(a[last = pos++]);
        }
        blackhole.consume(pos);
        blackhole.consume(last);
    }

    @Benchmark
    public void optimize(Blackhole blackhole) {
        pos = max >> 1;
        last = pos - 1;

        for (int i = pos; i < max; i++) {
            blackhole.consume(a[i]);
        }
        pos = max;
        last = max - 1;
        blackhole.consume(pos);
        blackhole.consume(last);
    }
}
```
Output:
```
Benchmark                            (max)  Mode  Cnt     Score    Error  Units
ForEachRemainingBenchmark.optimize      10  avgt    5     3.331 ±  0.018  ns/op
ForEachRemainingBenchmark.optimize    1000  avgt    5    56.202 ±  0.023  ns/op
ForEachRemainingBenchmark.optimize  100000  avgt    5  6028.862 ±  7.591  ns/op
ForEachRemainingBenchmark.raw           10  avgt    5     3.786 ±  0.006  ns/op
ForEachRemainingBenchmark.raw         1000  avgt    5    53.319 ±  0.026  ns/op
ForEachRemainingBenchmark.raw       100000  avgt    5  6007.510 ± 14.774  ns/op
```

Reply:
Well, that's a sensible criterion, but the ArrayList code has been written by many people over the years, and optimized for different implementations of Collections and different JVMs, so it's not necessary the best.

However, the difference is so small that I'd keep the new ArrayList-style code. It's definitely cleaner, and writing in Rust I realized that exceeding with side-effects in expression evaluation doesn't do a good service to anybody.